src/metabase/upload.clj

```diff
@@ -45,33 +45,55 @@
 ;;;; | Schema detection |
 ;;;; +------------------+
 
-;; Upload types form a DAG (directed acyclic graph) where each type can be coerced into any
-;; of its ancestors types. We parse each value in the CSV file to the most-specific possible
-;; type for each column. The most-specific possible type for a column is the lowest common
-;; ancestor of the types for each value in the column.
+;; Upload value-types form a DAG (directed acyclic graph) where each type can be coerced into any of
+;; its ancestors types. We parse each value in the CSV file to the most-specific possible type for
+;; each column. The most-specific possible type for a column is the closest common ancestor of the
+;; types for each value in the column, found by walking through the graph in topological order,
+;; following edges from left to right - not that this is not guaranteed to be a least common ancestor!
 ;;
-;;              text
-;;               |
-;;               |
-;;          varchar-255┐
-;;        /     / \    │
-;;       /     /   \   └──────────┬
-;;      /     /     \             │
-;;  boolean float   datetime  offset-datetime
-;;     |     │       │
-;;     │     │       │
-;;     |    int     date
-;;     |   /   \
-;;     |  /     \
-;;     | /       \
-;;     |/         \
-;; boolean-or-int  auto-incrementing-int-pk
+;; See [[metabase.util.ordered-hierarchy/first-common-ancestor]] for more details.
 ;;
-;; `boolean-or-int` is a special type with two parents, where we parse it as a boolean if the whole
-;; column's values are of that type. additionally a column cannot have a boolean-or-int type, but
-;; a value can. if there is a column with a boolean-or-int value and an integer value, the column will be int
-;; if there is a column with a boolean-or-int value and a boolean value, the column will be boolean
-;; if there is a column with only boolean-or-int values, the column will be parsed as if it were boolean
+;;                 text
+;;                  |
+;;                  │
+;;              varchar-255────────┐
```

Comment:
I was opening my mouth to say "this graph is becoming a maintenance nightmare", but the hairier it gets the gladder I am it exists…

Reply:
I'm fairly sure it used to have `<pre><code>` tags around it so that it would render right in Marginalia-land. We should get those back. (Or if I'm misremembering, add them)

Reply:
Oh, just saw the closing tags on L98. They should only surround the DAG, not the text, IMHO